`Tooltip` component (HDS-1743) (#1283)
hashicorp/design-system: website/docs/components/tooltip/partials/code/how-to-use.md

## How to use this component

The HDS tooltip is provided as both a `TooltipButton` component and as an `hds-tooltip` Ember [modifier](/components/tooltip?tab=code#ember-modifier). To be accessible, the element which triggers a tooltip needs to be an interactive element which can receive focus. This means that:

* if the tooltip needs to be applied to a non-interactive element, the `TooltipButton` component should be used as it wraps the opener element with a `button` element.
* if the tooltip needs to be applied to an element which is already interactive, the `hds-tooltip` modifier should be used.

Currently, the tooltip uses [Tippy.js](https://atomiks.github.io/tippyjs/) under the hood for its main functionality.

<!-- use the same heading order from Guidelines -->
### TooltipButton

!!! Warning

As the `TooltipButton` component wraps its content with an HTML `<button>` element, there may be possible layout changes when using it to wrap an existing non-interactive element in your application’s UI.
!!!

#### Icon
```handlebars
  <Hds::TooltipButton @text="Hello!">

>     <FlightIcon @name="info" aria-label="More information" />

Reviewer: ```suggestion
    <FlightIcon @name="info" aria-label="More information" />
```

@KristinLBradley the aria-label should be on the button, not the FlightIcon.

Author: Updated, thanks!

Use the `@placement` argument if you would like to use a different starting position for the tooltip vs. the default.

```handlebars
  <Hds::TooltipButton @text="Hello!" @placement="right">
    <Hds::Badge @text="Some tests failed" @icon="alert-triangle" @color="warning" />
  </Hds::TooltipButton>
```

#### Offset

You can change the offset of the tooltip in relation to the opener element content if needed.

```handlebars
  <Hds::TooltipButton @text="Hello!" @offset={{array 50 30}}>
    <Hds::Stepper::Step::Indicator @text="1" @status="incomplete" />
  </Hds::TooltipButton>
```

#### Extra Tippy Options

You can use `@extraTippyOptions` to provide more specific options to [Tippy.js](https://atomiks.github.io/tippyjs/). For a full list of available options refer to the [Tippy.js API documentation](https://atomiks.github.io/tippyjs/v6/all-props/).

For example, you can use the `allowHTML` option to enable rich tooltip text:

```handlebars
  <Hds::TooltipButton
    @extraTippyOptions={{hash allowHTML=true}}

>     @text="<b>Hello</b> <em>there</em>! Here is more info."

Author: ```suggestion
    @text="<b>Hello</b> <em>there</em>!"
```

<Hds::TooltipButton @text="Hello!" aria-label="More information">
  <FlightIcon @name="info" />
</Hds::TooltipButton>
```

#### Inline with text
```handlebars
<h4 class="hds-typography-display-100">
  Header text
  <Hds::TooltipButton @text="Hello!" aria-label="More information">
    <FlightIcon @name="info" />
  </Hds::TooltipButton>
</h4>
```

#### Placement

The tooltip appears at the “top” centered above the opener button content by default. If the tooltip is near any of the edges of the screen, its position will automatically adjust to prevent the tooltip content from being cut off so it remains readable.

Use the `@placement` argument if you would like to use a different starting position for the tooltip vs. the default.

```handlebars
<Hds::TooltipButton @text="Hello!" @placement="right">
  <Hds::Badge @text="Some tests failed" @icon="alert-triangle" @color="warning" />
</Hds::TooltipButton>
```

#### Offset

You can change the offset of the tooltip in relation to the opener element content if needed.

```handlebars
<Hds::TooltipButton @text="Hello!" @offset={{array 50 30}}>
  <Hds::Stepper::Step::Indicator @text="1" @status="incomplete" />
</Hds::TooltipButton>
```

> ### isInline

Reviewer 2: ```suggestion
#### isInline
```

You can change the default `inline-block` display to use `flex` display if needed.

```handlebars
<p>
  Text before
  <Hds::TooltipButton 
    @text="Hello!" 
    aria-label="HashiCorp" 
    @isInline={{false}}
  >
    <FlightIcon @name="hashicorp" />
  </Hds::TooltipButton>
  Text after
</p>
```

#### Extra Tippy Options

You can use `@extraTippyOptions` to provide more specific options to [Tippy.js](https://atomiks.github.io/tippyjs/). For a full list of available options refer to the [Tippy.js API documentation](https://atomiks.github.io/tippyjs/v6/all-props/).

For example, you can use the `allowHTML` option to enable rich tooltip text:

```handlebars
<Hds::TooltipButton
  @extraTippyOptions={{hash allowHTML=true}}
  @text="<b>Hello</b> <em>there</em>!"
>
  More information
</Hds::TooltipButton>
```

!!! Warning

If you enable the `allowHTML` option:
- be sure to sanitize your data
- to maintain accessibility, do not include interactive content such as links or buttons
- we recommend using only basic inline-level text formatting tags such as `strong` or `em`; using block-level tags such as `div` or `p` will make the HTML syntax invalid

!!!

### Ember modifier

> An Ember modifier is available if your use case requires attaching a tooltip to an element than is already interactive (to be accessible, tooltips should only be attached to interactive elements, like buttons, links, inputs, etc).

Reviewer 3: ```suggestion
An Ember modifier is available if your use case requires attaching a tooltip to an element that is already interactive (to be accessible, tooltips should only be attached to interactive elements, like buttons, links, inputs, etc).
```

#### Modifier used on a link

```handlebars
<Hds::Link::Standalone 
  {{hds-tooltip "Hello!"}} 
  @href="#" 
  @icon="collections" 
  @text="Read tutorial" 
/>
```

#### Placement
```handlebars
<Hds::Button 
  {{hds-tooltip "Hello!" 
  options=(hash placement="right")}} 
  @icon="external-link" 
  @text="Visit website" 
  @href="https://hashicorp.com" 
/>
```

#### Offset
```handlebars
<Hds::Form::Field @layout="vertical" as |F|>
  <F.Label @controlId="tooltip-example-control-id">First Name</F.Label>
  <F.Control>
    <Hds::Form::TextInput::Base 
      {{hds-tooltip "Hello!" 
      options=(hash placement="top" 
      offset=(array 0 30))}} 
      @type="text" 
      @value="Jane" 
      id="tooltip-example-control-id"
      @width="200px" 
    />
  </F.Control>
</Hds::Form::Field>
```

#### Extra Tippy Options

> You can enable extra [Tippy.js options](https://atomiks.github.io/tippyjs/v6/all-props/) by passing a hash of the options you wish to use similarly to how the `TooltipButton` component works.

Reviewer: I wonder if this would be the right place to caution consumers about the potential non-conformance of some tippy options.

```suggestion
While you can enable extra [Tippy.js options](https://atomiks.github.io/tippyjs/v6/all-props/) by passing a hash of the options you wish to use similarly to how the `TooltipButton` component works, you should make sure that any additional options chosen will still result in a WCAG-conformant outcome.
```

Reviewer 2: Yes, but I would not start the paragraph with `While you can`.  We want to convey that they **can**. And then specify that in that case they have to make sure the result is WCAG-conformant.

Author: How about:

You can enable extra [Tippy.js options](https://atomiks.github.io/tippyjs/v6/all-props/) by passing a hash of the options you wish to use similarly to how the `TooltipButton` component works. You should make sure that any additional options chosen will still result in a WCAG-conformant outcome.

Author: I went ahead and updated the text as I suggested above. Please open another comment if you think it still needs changing.

> You can enable extra [Tippy.js options](https://atomiks.github.io/tippyjs/v6/all-props/) by passing a hash of the options you wish to use similarly to how the `TooltipButton` component works.

Author: ```suggestion
You can enable extra [Tippy.js options](https://atomiks.github.io/tippyjs/v6/all-props/) by passing a hash of the options you wish to use similarly to how the `TooltipButton` component works. You should make sure that any additional options chosen will still result in a WCAG-conformant outcome.
```

## How to use this component

The HDS tooltip is provided both as a `TooltipButton` component and as a `tooltip` Ember modifier. The component is intended to handle the majority of use cases and should be used vs. the modifier whenever feasible. The modifier is available if you need to apply a tooltip to a custom element other than an HTML button although should still only be used with interactive elements such as anchor links and form inputs to ensure accessibility.

> Currently the tooltip uses [https://atomiks.github.io/tippyjs/](Tippy.js) under the hood for its main functionality.

Author: ```suggestion
Currently the tooltip uses [Tippy.js](https://atomiks.github.io/tippyjs/) under the hood for its main functionality.
```

> Currently the tooltip uses [https://atomiks.github.io/tippyjs/](Tippy.js) under the hood for the majority of its functionality.

Author: ```suggestion
Currently the tooltip uses [https://atomiks.github.io/tippyjs/](Tippy.js) under the hood for its main functionality.
```

<!-- use the same heading order from Guidelines -->
### Basic examples 

#### With FlightIcon
```handlebars
  <Hds::TooltipButton @text="Here is more information">
    <FlightIcon @name="alert-circle" aria-label="More information" />
  </Hds::TooltipButton>
```

#### With text content and tooltip text formatted with bold and em tags
```handlebars
  <Hds::TooltipButton @text="<b>Hello</b> <em>there</em>! Here is more info.">
    More information
  </Hds::TooltipButton>
```

> #### Used next to text content

Author: I used it within and next to text in the showcase to see how it looks and test for any major layout issues. I'm not sure why I would add a flexbox parent.

If we want to include an explanation to the consumer's about vertical alignment, where in the docs should I add that?

```handlebars
  <p>
    <strong>Header text</strong>
    <Hds::TooltipButton @text="Here is more information">
      <FlightIcon @name="alert-circle" aria-label="More information" />
    </Hds::TooltipButton>
  </p>
```

### Placement

The tooltip appears at the “top” centered above the opener button content by default. If the tooltip is near any of the edges of the screen, its position will automatically adjust to prevent the tooltip content from being cut off so it remains readable.

> Use the `placement` argument if you would like to use a different starting position for the tooltip vs. the default.

Author: I guess including "@" for "@placement" could help make it clearer.

> Use the `placement` argument if you would like to use a different starting position for the tooltip vs. the default.

Author: I'm going to go ahead and add it.

## How to use this component

The HDS tooltip is provided as both a `TooltipButton` component and as an `hds-tooltip` Ember modifier. To be accessible, the element which triggers a tooltip needs to be an interactive element which can receive focus. This means that:

* if the tooltip needs to be applied to a non-interactive element, the `TooltipButton` component should be used as it wraps the opener element with a `button` element.

> * if the tooltip needs to be applied to an element which is already interactive, the `hds-tooltip` mmodifier should be used.

Reviewer 2: ```suggestion
* if the tooltip needs to be applied to an element which is already interactive, the `hds-tooltip` modifier should be used.
```

Currently, the tooltip uses [Tippy.js](https://atomiks.github.io/tippyjs/) under the hood for its main functionality.

<!-- use the same heading order from Guidelines -->

> ### Content

Reviewer 2: ```suggestion
### TooltipButton
```

  </Hds::TooltipButton>
```

#### Inline with text
```handlebars
  <p>
    <strong>Header text</strong>
    <Hds::TooltipButton @text="Here is more information">
      <FlightIcon @name="alert-circle" aria-label="More information" />
    </Hds::TooltipButton>
  </p>
```

#### Placement

The tooltip appears at the “top” centered above the opener button content by default. If the tooltip is near any of the edges of the screen, its position will automatically adjust to prevent the tooltip content from being cut off so it remains readable.

Use the `@placement` argument if you would like to use a different starting position for the tooltip vs. the default.

```handlebars
  <Hds::TooltipButton @text="Here is more information" @placement="bottom">
    <FlightIcon @name="alert-circle" aria-label="More information" />
  </Hds::TooltipButton>
```

#### Offset

You can change the offset of the tooltip in relation to the opener element content if needed.

```handlebars
  <Hds::TooltipButton @text="Here is more information" @offset={{array 50 30}}>
    <FlightIcon @name="alert-circle" aria-label="More information" />
  </Hds::TooltipButton>
```

#### Extra Tippy Options

> Using `@extraTippyOptions` to enable rich tooltip text.

Reviewer 2: ```suggestion
You can use `@extraTippyOptions` to enable rich tooltip text.
```

Reviewer 2: Also, probably this sub-section should be expanded a bit: `@extraTippyOptions` _can_ be used to enable rich text, but also for other things (essentially providing more options to Tippy.js.

We should be clear on this, and probably even link to the [All Props](https://atomiks.github.io/tippyjs/v6/all-props/) page of Tippy.

I would imagine something like this:
- explain what `@extraTippyOptions` can do/is useful for
- explain that we can't document all the possible options, so they should refer to the official API documentation for details
- explain that this can be used to enable rich text
- show a demo code snippet of using `allowHTML`
- explain that sanitization of the code is responsibility of the consumers

Be sure to sanitize your data if enabling this option. To maintain accessibility, do not include interactive content such as links or buttons. We recommend using only basic inline-level text formatting tags such as `strong` or `em`. Using block-level tags such as `div` or `p` will make the HTML syntax invalid.

```handlebars
  <Hds::TooltipButton 
    @extraTippyOptions={{hash allowHTML=true}}
    @text="<b>Hello</b> <em>there</em>! Here is more info."
  >
    More information
  </Hds::TooltipButton>
```

### Ember modifier

> An Ember modifier is available if your use case requires attaching a tooltip to an element other than a button. To be accessible, tooltips should only be attached to interactive elements. You will need to add your own styling for the element you are attaching the toolip to.

Reviewer 2: ```suggestion
An Ember modifier is available if your use case requires attaching a tooltip to an element than is already interactive (to be accessible, tooltips should only be attached to interactive elements, like buttons, links, inputs, etc). 
```

#### Modifier used on a link

```handlebars

>   <a href="#" {{hds-tooltip "This link takes you to more information"}}>More information</a>

Reviewer 2: also for the other examples, so the options are visible without scrolling? otherwise all the code snippets look exactly the same (unless one scrolls horizontally the code snippet)

```suggestion
  <a href="#" {{hds-tooltip "Hello"}}>More information</a>
```

```

> #### Placement option

Reviewer 2: ```suggestion
#### Placement
```

```handlebars
  <a href="#" {{hds-tooltip "This link takes you to more information" options=(hash placement="right")}}>More information</a>
```

> #### Offset option

Reviewer 2: ```suggestion
#### Offset
```

```handlebars
  <a href="#" {{hds-tooltip "This link takes you to more information" options=(hash offset=(array 60 60))}}>More information</a>
```

> 

Reviewer 2: should we mention here something about the fact that the modifier accepts (I think) also other options, similar to what the `extraTippyOptions` does when using the "component" variant?
